### DISCUSSION TITLE
Subclassing geodataframe

### DISCUSSION THREAD

**Discussion by awa5114** [geopandas/geopandas#2750] [Q&A]
What is the best way to subclass a `GeoDataFrame`? I tried to follow some instructions as indicated in [this link](https://pandas.pydata.org/pandas-docs/stable/development/extending.html#subclassing-pandas-data-structures) for subclassing pandas data structures but this does not seem to propagate to the `from_file` classmethod:

```
import geopandas

class Hectometer(geopandas.GeoSeries):
    #each instance represents 1 hectometer
    @property
    def _constructor(self):
        return Hectometer

    @property
    def _constructor_expanddim(self):
        return Hectometers

class Hectometers(geopandas.GeoDataFrame):
    @property
    def _constructor(self):
        return Hectometers

    @property
    def _constructor_sliced(self):
        return Hectometer
def test():
    hms = Hectometers.from_file(r'vector\Hectometerraai_A2.shp')
    print(type(hms))
if __name__ == '__main__':
    test()
```

The above code outputs `<class 'geopandas.geodataframe.GeoDataFrame'>`, whereas I would like to get the subclassed object instead. I tried about overriding the `from_file` constructor but this does not seem to refer to `self` in the original `GeoDataFrame` class which means it will always return a `GeoDataFrame` and not its subclass.

What is an elegant way to subclass `GeoDataFrames` and `GeoSeries` such that I retrieve the subclassed object from the `from_file` method?

**Comment by awa5114** [comment]
I think the problem here is that the factory methods `from_file` and `from_features` are not referring to the `cls` argument to do their work. Is there any reason for this? It is highly limiting for users who wish to subclass a `GeoDataFrame`.... Please take a look at my [SE post ](https://stackoverflow.com/questions/63260633/subclassing-a-class-from-geopandas)which explores some other aspects of this problem in more detail.

**Comment by jorisvandenbossche** [comment]
@awa5114 sorry for the late reply

In general, subclassing GeoDataFrame is not really supported: we (as geopandas project) never really considered this use case or made effort in making this easier / testing this. 
In pandas itself, there is some functionality to enable subclassing (like the `_constructor`, and `__finalize__` methods), and which geopandas makes use of to implement its DataFrame subclass. But we (currently) don't use it in a way that makes further subclassing easy. 
(see also some discussion about this in https://github.com/CartoDB/cartoframes/pull/1130#issuecomment-557695873, where a subclass was also considered / tried).

It's certainly possible for GeoPandas to update its code to make better use of things like `_constructor` so your code example from above would work. But such features also come with a maintenance cost of course. 
Can I ask what your use case is to subclass? Since GeoPandas uses a subclass itself, I certainly understand the value of a subclass, but nowadays, pandas also provides several other ways to add functionality to dataframes without subclassing (eg registering custom accessors, adding metadata in `.attrs`, ..)

**Comment by awa5114** [comment]
I don't know if my specific use case warrants such a high maintenance cost. It is very specific personal code that I'm using for educational purposes. I'll take a look at the other alternatives you proposed...

**Comment by NickLucche** [comment]
Hey, despite @jorisvandenbossche absolutely legitimate and functional warning against using a custom geodataframe subclass, my experience in using it for quite some months now at work is that it comes extremely in handy as it is such a familiar concept to interact with for programmers. 
In particular, new people coming in are able to quickly get a grasp of it by simply being acquainted with pandas API, while people that are totally new to the language can simply follow along geopandas tutorial while enjoying some extra checks/warning that I've added in multiple parts which highlight some of the most common mistakes that are either general to geodata or very specific to our needs. It also helps in having a "lightweight support" for having multiple (non-active) geometry columns with a synchronized crs.  

I believe that it would be a very nice feature to have in geopandas and I understand that I've barely scratched the surface of the problem with the PR https://github.com/geopandas/geopandas/pull/2433. Nonetheless, I'd be glad to put in some more effort to come up with a solution that fits your project ideas.  

Currently, my own implementation of a custom subclass (aka `Dataset`) relies on some workarounds to keep the `Dataset` class around after operations are performed on it. Aside from implementing the `_constructor`, I re-wrapped all methods I needed in which geopandas specifically set the `__class__` property (essentially doing the same thing):
```python
  def merge(self, *args, **kwargs):
      # this must be explicitely overwritten as geopandas modifies __class__ in `merge`
      res = super().merge(*args, **kwargs)
      if isinstance(res, gpd.GeoDataFrame):
          res.__class__ = Dataset
      return res

    def __getitem__(self, key):
        res = super().__getitem__(key)
        geo_col = self._geometry_column_name
        if isinstance(res, gpd.GeoDataFrame) and geo_col in res:
            res.__class__ = Dataset
            res._geometry_column_name = geo_col
        return res
``` 

My suggestion is that we start with the most obvious parts by addressing the `from_*` classmethods, I don't really see any downsides in using the `cls` constructor instead of specifically setting `GeoDataFrame`, as this would be a fairly low-effort change.
I have some ideas on making at least the `__class__` property set "dynamic", but I would appreciate to hear your opinion on the matter in the bigger pandas/geopandas picture.

**Comment by m-richards** [comment]
@NickLucche I don't have the same background as Joris to comment within the context of the pandas subclassing. I would also just reiterate the point about the use case to subclass - and whether what you're doing with subclassing a geodataframe could be achived through composition of a Dataset class containing a geodataframe as an attribute (but if you've been using it for months, it's quite likely you already have a good reason). 

From my perspective, I see changing the `from_*` classmethods as an innocuous change - it would at least help your use case. But there would still be a considerable amount of work to have properly implemented subclassing - because every time `_constructor` is called internally by pandas, the custom subclass would be lost (which as an aside if you've been developing your class wrapper with geopandas 0.10.2, `_constructor` has changed on the main branch and would be more complicated to feed a class through.

Also there would be a considerable testing burden to cover all cases and noting replacing GeoDataFrame / GeoSeries with `cls` in every spot will not always do the desired thing (for instance GeoDataFrame.set_geometry is monkey patched onto pd.DataFrame, so if we replace the GeoDataFrame constructor call within, then it will no longer produce GeoDataFrames, but rather dataframes).

**Comment by NickLucche** [comment]
Hey @m-richards, thanks a lot for your response and your insights! My implementation is indeed based on 0.10.0 and it relies on a custom `_constructor`to create instances, as suggested (at the time) by the pandas documentation, and I would favor maintaining this behavior of having to provide a custom `_constructor` to make a subclass.
But I see your point in the efforts required to having such changes tested on a feature you didn't intend to support, so I won't move forward until testing out a few implementations options based on your feedback.

Nonetheless, I would personally find it useful to have the tiny changes proposed in PR #2433 merged in the main branch, let me know if you have further feedback on that.
